cpu/native/fs/native_fs.c

```diff
@@ -0,0 +1,295 @@
+{
+    static char buffer[PATH_MAX];
+
+    _do_prefix(mountp, name, buffer, sizeof(buffer));
+
+    return buffer;
+}
+
+static int _mount(vfs_mount_t *mountp)
+{
+    int res;
+    mutex_lock(&_lock);
+
+    /* create common root dir first */
+    if (CONFIG_NATIVE_ISOLATE_FS) {
+        char *parent = _prefix_path(mountp, "");
+        /* remove node specific suffix */
+        for (unsigned i = 0; i < 2; ++i) {
+            char *end = strrchr(parent, '/');
+            *end = '\0';
+        }
+        real_mkdir(parent, 0777);
+    }
+
+    real_mkdir(_prefix_path(mountp, ""), 0777);
+    res = errno == EEXIST ? 0 : -errno;
+
+    mutex_unlock(&_lock);
+    return res;
+}
+
+static int _unmount(vfs_mount_t *mountp)
+{
+    mutex_lock(&_lock);
+
+    real_rmdir(_prefix_path(mountp, ""));
+
+    mutex_unlock(&_lock);
+    return 0;
+}
+
+static int _mkdir(vfs_mount_t *mountp, const char *name, mode_t mode)
+{
+    int res = 0;
+    mutex_lock(&_lock);
+
+    if (real_mkdir(_prefix_path(mountp, name), mode) < 0) {
+        res = -errno;
+    }
+
+    mutex_unlock(&_lock);
+    return res;
+}
+
+static int _rmdir(vfs_mount_t *mountp, const char *name)
+{
+    int res = 0;
+    mutex_lock(&_lock);
+
+    if (real_rmdir(_prefix_path(mountp, name)) < 0) {
+        res = -errno;
+    }
+
+    mutex_unlock(&_lock);
+    return res;
+}
+
+static int _statvfs(vfs_mount_t *mountp, const char *restrict path, struct statvfs *restrict buf)
+{
+    int res = 0;
+    mutex_lock(&_lock);
+
+    if (real_statvfs(_prefix_path(mountp, path), buf) < 0) {
+        res = -errno;
+    }
+
+    mutex_unlock(&_lock);
+    return res;
+}
+
+static int _open(vfs_file_t *filp, const char *name, int flags, mode_t mode)
+{
+    int res = 0;
+    mutex_lock(&_lock);
+
+    if ((FD(filep) = real_open(_prefix_path(filp->mp, name), flags, mode)) < 0) {
+        res = -errno;
+    }
+
+    mutex_unlock(&_lock);
+    return res;
+}
+
+static ssize_t _read(vfs_file_t *filp, void *dest, size_t nbytes)
+{
+    return real_read(FD(filep), dest, nbytes);
+}
+
+static ssize_t _write(vfs_file_t *filp, const void *src, size_t nbytes)
+{
+    return real_write(FD(filep), src, nbytes);
+}
+
+static off_t _lseek(vfs_file_t *filp, off_t off, int whence)
+{
+    return real_lseek(FD(filep), off, whence);
+}
+
+static int _fstat(vfs_file_t *filp, struct stat *buf)
+{
+    return real_fstat(FD(filep), buf);
+}
+
+static int _fsync(vfs_file_t *filp)
+{
+    return real_fsync(FD(filep));
+}
+
+static int _close(vfs_file_t *filp)
+{
+    return real_close(FD(filep));
+}
+
+static int _unlink(vfs_mount_t *mountp, const char *name)
+{
+    int res = 0;
+    mutex_lock(&_lock);
+
+    if (real_unlink(_prefix_path(mountp, name)) < 0) {
+        res = -errno;
+    }
+
+    mutex_unlock(&_lock);
+    return res;
+}
+
+static int _rename(vfs_mount_t *mountp, const char *from_path, const char *to_path)
+{
+    static char path_a[PATH_MAX];
```

Comment:
I wonder whether it's warranted to have two PATH_MAX buffers around for an operation that can't be GC'd at link time because it's behind the file system driver -- but it's OK, we're on native, nothing in here will eat significant resources of the host anyway.

(If this were targetting embedded, I'd suggest to have static buffers per compilation unit and use them as needed).